### PR TITLE
[layout][bugfix] Horizontal & Vertical layouts shouldn't display children that have `display: none`

### LINKED ISSUE #394: `display: none` is being ignored in vertical and horizontal layouts
Example app:

```python
from textual.app import App
from textual.widget import Widget

class ButtonsApp(App):

    def on_load(self) -> None:
        self.bind("q", "quit")

    def on_mount(self):
        self.mount(
            Widget(classes={"no-display"})
        )


ButtonsApp.run(css_file="buttons.scss", log="textual.log")
```

Example CSS:
```scss
Widget.no-display {
  display: none;
}
```

The widget is unexpectedly displayed.

### PR DESCRIPTION
**Pull request by DrBenton**
fixes #394

### CODE REVIEW

File: src/textual/_node_list.py [hunk 1]
```diff
@@ -18,8 +18,8 @@ class NodeList:
 
     """
 
-    def __init__(self) -> None:
-        self._nodes: list[DOMNode] = []
+    def __init__(self, nodes: list[DOMNode] = None) -> None:
```
Comment (DrBenton) [inline src/textual/_node_list.py:21]: I was thinking that being able to optionally provide an initial list of nodes to this constructor could be useful?
I can revert this change if you think it's better to not do this of course :slightly_smiling_face: 

(I checked the [NodeList](https://devdocs.io/dom/nodelist) Web API, but there is no constructor on this one anyway so we cannot do any features comparison here :sweat_smile: )

File: src/textual/app.py [hunk 2]
```diff
@@ -521,6 +521,7 @@ async def process_messages(self) -> None:
                 mount_event = events.Mount(sender=self)
                 await self.dispatch_message(mount_event)
 
+                # TODO: don't override `self.console` here
```
Comment (DrBenton) [inline src/textual/app.py:524]: as discussed together :-)

File: src/textual/css/styles.py [hunk 3]
```diff
@@ -70,7 +70,7 @@
 class RulesMap(TypedDict, total=False):
     """A typed dict for CSS rules.
 
-    Any key may be absent, indiciating that rule has not been set.
+    Any key may be absent, indicating that rule has not been set.
```
Comment (DrBenton) [inline src/textual/css/styles.py:73]: [while I was there] As I was digging into the code of Textual I noticed this typo, highlighted by PyCharm

File: src/textual/layouts/horizontal.py [hunk 4]
```diff
@@ -39,7 +39,9 @@ def arrange(
 
         x = box_models[0].margin.left if box_models else 0
 
-        for widget, box_model, margin in zip(parent.children, box_models, margins):
+        displayed_children = parent.displayed_children
```
Comment (DrBenton) [inline src/textual/layouts/horizontal.py:42]: Contrary to `parent.children`, `parent.displayed_children` is a Python property getter; with this alias we'll trigger this property getter only once, and also makes the following code a wee bit easier to read

File: tests/layouts/test_common_layout_features.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+@pytest.mark.parametrize(
+    "layout,display,expected_in_displayed_children",
+    [
+        ("dock", "block", True),
+        ("horizontal", "block", True),
+        ("vertical", "block", True),
+        ("dock", "none", False),
+        ("horizontal", "none", False),
+        ("vertical", "none", False),
```
Comment (DrBenton) [inline tests/layouts/test_common_layout_features.py:15]: I noticed the existence of a `GridLayout`, but it's not managed in the layouts factory so I reckon it's a Work In Progress API?

Comment (darrenburns) [reply tests/layouts/test_common_layout_features.py:15]: It's an older layout that I think we're scrapping

File: src/textual/dom.py [hunk 6]
```diff
@@ -278,6 +278,10 @@ def add_children(tree, node):
         add_children(tree, self)
         return tree
 
+    @property
+    def displayed_children(self) -> NodeList:
```
Comment (willmcgugan) [inline src/textual/dom.py:282]: I don't think this needs to be a `NodeList`, which is used to control how the dev adds widgets. Plain old List is fine.

File: src/textual/_node_list.py (hunk 1 above)
Comment (willmcgugan) [reply src/textual/_node_list.py:21]: Not required ATM

**Review by willmcgugan — CHANGES_REQUESTED** [review]

**Comment by DrBenton** [comment]
@willmcgugan Thanks for your feedback! :slightly_smiling_face: 
--> addressed there: [72555f4](https://github.com/Textualize/textual/pull/436/commits/72555f4a77b594c21125b170d4dae75433db8955) (:crossed_fingers:)

**Review by willmcgugan — APPROVED** [review]

**Pull request merged**